jolt-core/src/zkvm/prover.rs

```diff
@@ -0,0 +1,1420 @@
+        SumcheckInstanceProof<F, ProofTranscript>,
+    ) {
+        #[cfg(not(target_arch = "wasm32"))]
+        print_current_memory_usage("Stage 2 baseline");
+        let span = tracing::span!(tracing::Level::INFO, "Stage 2 sumchecks");
+        let _guard = span.enter();
+
+        // Stage 2a: Prove univariate-skip first round for product virtualization
+        let (uni_skip_state, first_round_proof) = prove_stage2_uni_skip(
+            &self.trace,
+            &self.opening_accumulator,
+            &self.spartan_key,
+            &mut self.transcript,
+        );
+
+        let mut spartan_inner = InnerSumcheckProver::gen(
+            &self.opening_accumulator,
+            &self.spartan_key,
+            &mut self.transcript,
+        );
+        let mut spartan_product_virtual_remainder =
+            ProductVirtualRemainderProver::gen(&self.trace, &uni_skip_state);
+        let mut ram_raf_evaluation = RamRafEvaluationSumcheckProver::gen(
+            &self.trace,
+            &self.program_io.memory_layout,
+            self.ram_K,
+            &self.opening_accumulator,
+        );
+        let mut ram_read_write_checking = RamReadWriteCheckingProver::gen(
+            &self.initial_ram_state,
+            &self.preprocessing.bytecode,
+            &self.program_io.memory_layout,
+            &self.trace,
+            self.ram_K,
+            self.twist_sumcheck_switch_index,
+            &self.opening_accumulator,
+            &mut self.transcript,
+        );
+        let mut ram_output_check = OutputSumcheckProver::gen(
+            &self.initial_ram_state,
+            &self.final_ram_state,
+            &self.program_io,
+            self.ram_K,
+            &mut self.transcript,
+        );
+
+        #[cfg(feature = "allocative")]
+        {
+            let mut flamegraph = FlameGraphBuilder::default();
+            for sumcheck in stage2_instances.iter() {
+                sumcheck.update_flamegraph(&mut flamegraph);
+            }
+            write_flamegraph_svg(flamegraph, "stage2_start_flamechart.svg");
+        }
+
+        tracing::info!("Stage 2 proving");
+        let (sumcheck_proof, _r_stage2) = BatchedSumcheck::prove(
+            vec![
+                &mut spartan_inner,
+                &mut spartan_product_virtual_remainder,
+                &mut ram_raf_evaluation,
+                &mut ram_read_write_checking,
+                &mut ram_output_check,
+            ],
+            &mut self.opening_accumulator,
+            &mut self.transcript,
+        );
+
+        #[cfg(feature = "allocative")]
+        {
+            let mut flamegraph = FlameGraphBuilder::default();
+            for sumcheck in stage2_instances.iter() {
+                sumcheck.update_flamegraph(&mut flamegraph);
+            }
+            write_flamegraph_svg(flamegraph, "stage2_end_flamechart.svg");
+        }
+
+        // Drop product immediately because it has lifetimes.
+        drop(spartan_product_virtual_remainder);
```

Comment:
I slightly prefer to keep trace as an `Arc<Vec<Cycle>>` instead introducing the lifetime, so that we can use `drop_in_background` for everything. Same for the shift sumcheck.